x509/ekcert_test.go

```diff
@@ -0,0 +1,446 @@
+		},
+		{
+			name: "HasEkcExtendedKeyUsage is false",
+			modifyCertTemplate: func(t *testing.T, cert *x509.Certificate) {
+				cert.UnknownExtKeyUsage = []asn1.ObjectIdentifier{}
+			},
+			wantHasEkcExtendedKeyUsage: false,
+		},
+		{
+			name: "Unhandled critical extension",
+			modifyCertTemplate: func(t *testing.T, cert *x509.Certificate) {
+				// Use empty subject, make the SubjectAltName extension critical, and
+				// add an unhandled critical extension.
+				cert.Subject = pkix.Name{}
+				for i, ext := range cert.ExtraExtensions {
+					if ext.Id.Equal(oid.SubjectAltName) {
+						cert.ExtraExtensions[i].Critical = true
+					}
+				}
+				cert.ExtraExtensions = append(cert.ExtraExtensions, pkix.Extension{
+					Id:       asn1.ObjectIdentifier{1, 2, 3, 4, 5},
+					Critical: true,
+					Value:    []byte("test"),
+				})
+			},
+			wantHasEkcExtendedKeyUsage:          true,
+			wantUnhandledCriticalExtensionsOIDs: []asn1.ObjectIdentifier{{1, 2, 3, 4, 5}},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			cert := setupTestCert(t, test.modifyCertTemplate)
+			ekCert, err := ToEKCertificate(cert)
+			if err != nil {
+				t.Fatalf("ToEKCertificate() returned an unexpected error: %v", err)
+			}
+
+			if ekCert.Certificate != cert {
+				t.Errorf("ToEKCertificate() Certificate = %v, want: %v", ekCert.Certificate, cert)
+			}
+
+			if ekCert.TpmManufacturer != SAN.DirectoryNames[0].ExtraNames[0].Value {
+				t.Errorf("ToEKCertificate() TPMManufacturer = %v, want: %v", ekCert.TpmManufacturer, SAN.DirectoryNames[0].ExtraNames[0].Value)
+			}
+
+			if ekCert.TpmModel != SAN.DirectoryNames[0].ExtraNames[1].Value {
+				t.Errorf("ToEKCertificate() TPMModel = %v, want: %v", ekCert.TpmModel, SAN.DirectoryNames[0].ExtraNames[1].Value)
+			}
+
+			if ekCert.TpmVersion != SAN.DirectoryNames[0].ExtraNames[2].Value {
+				t.Errorf("ToEKCertificate() TPMVersion = %v, want: %v", ekCert.TpmVersion, SAN.DirectoryNames[0].ExtraNames[2].Value)
+			}
+
+			if ekCert.TpmSpecification != (tpmSpec) {
+				t.Errorf("ToEKCertificate() TpmSpecification is %v, want %v", ekCert.TpmSpecification, tpmSpec)
+			}
+
+			if test.wantHasEkcExtendedKeyUsage != ekCert.HasEkcExtendedKeyUsage {
+				t.Errorf("ToEKCertificate() HasEkcExtendedKeyUsage = %v, want: %v", ekCert.HasEkcExtendedKeyUsage, test.wantHasEkcExtendedKeyUsage)
+			}
+
+			sortOIDs := cmpopts.SortSlices(func(a, b asn1.ObjectIdentifier) bool { return a.String() < b.String() })
+			if diff := cmp.Diff(test.wantUnhandledCriticalExtensionsOIDs, ekCert.UnhandledCriticalExtensions, cmpopts.EquateEmpty(), sortOIDs); diff != "" {
+				t.Errorf("ToEKCertificate() UnhandledCriticalExtensions differs (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestToEKCertificate_FailuresTests(t *testing.T) {
+	tests := []struct {
+		name       string
+		modifyCert func(*testing.T, *x509.Certificate)
+		wantErr    error
+	}{
+		{
+			name: "Version is 2 (failure)",
```

Comment:
is ` (failure)` necessary? Each of these are tests nested under TestToEkCertificate_Failures